README.md
<h1 align="center">vidar</h1>
<div align="center">
<a href="https://www.npmjs.com/package/@raminjafary/vidar"><img src="https://badgen.net/npm/v/@raminjafary/vidar" alt="Version"></a>
</div>

<p align="center"><strong>Generate high-performance heatmap.</strong></p>

<p align="center">Fork of simpleheat.js with more focus on perfomance and maintainable code.</p>

[Demo](./playground/front)

## Usage

#### Rendering

```js
// create a heatmap object given an id or canvas reference
const heatmap = new Canvas2dRenderer(canvas);
// draw the heatmap
heatmap.draw();
```
#### Rate limit rendering 

You can throttle rendering per frames based on available resources. In the following example, the `draw` function will always be called with the latest data points with the frame rates browsers provide (e.g. 20 calls on 20fps).

```js
// throttle draw
const frame = ensureAvailableFrames(heatmap.draw)
// run on available frame, pass any args if you have
frame.request()
// cancel a frame
frame.cancel()
```

#### Data

```js
// set data of [[x, y, value], ...] format
heatmap.data = data

// add a data point
heatmap.addData(point);

// clear data
heatmap.clearData();
```

#### Appearance

```js
// set point radius and blur radius (25 and 15 by default)
heatmap.setRadius(radius, blur);

// set gradient colors as {<stop>: '<color>'}
// e.g. {0.4: 'blue', 0.65: 'lime', 1: 'red'}
heatmap.setGradient(gradient);

// call in case Canvas size changed
heatmap.resizeCanvas();
```

#### DOM Screenshot with [html2canvas](https://github.com/niklasvh/html2canvas)
```js
// you can take screenshot from DOM at a given time with data points baked in
const capture = document.querySelector('#capture')
html2canvas(capture)
  .then(canvas => {
    document.body.appendChild(canvas)
    canvas.style.display = 'none'
    return canvas
  })
  .then(canvas => {
    const image = canvas.toDataURL('image/png')
    // do whatever with data and then remove the canvas
    canvas.remove()
  })

```
## Development
- Clone this repository.
- Install dependencies using `yarn install` or `npm install`.
- Start development server using `yarn dev`.
- Follow the [Conventional Commits Specification](https://conventionalcommits.org) for opening PRs.

## License

[MIT](./LICENSE)



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2021-07-12)


### Features

* add  functionalities to get coordinations on all screen sizes ([a364a7d](https://github.com/raminjafary/vidar/commit/a364a7d00afe197a54a432df1b488c2f3158e602))
* add another example for the new implementation ([2abff66](https://github.com/raminjafary/vidar/commit/2abff664038d79e514e72516a9830a1aa7e260f7))
* add custom dimensions for heatmap ([00e9bcb](https://github.com/raminjafary/vidar/commit/00e9bcbf55f335450fb3cabe2dc26ea7f5208348))
* add examples for frontend ([c0bbd1b](https://github.com/raminjafary/vidar/commit/c0bbd1bc1fb1bf9504f20118d80c2fda2b305622))
* add new implementaion ([2f87c6e](https://github.com/raminjafary/vidar/commit/2f87c6e79f9432be28754c569d0815bfd2cca082))
* finalize new implementaion ([cfe92a1](https://github.com/raminjafary/vidar/commit/cfe92a1b99be03cd65fbdbb2106baf57459dad9a))
* generate heatmap based on coordmap ([2f1e041](https://github.com/raminjafary/vidar/commit/2f1e0416a2c83896b19a601cd2e52c9cf1bb099d))
* **utility:** add throttle with available frames fn ([ad1e029](https://github.com/raminjafary/vidar/commit/ad1e02919a6eb1e89b2d1d689ebce700cc3f8578))


### Bug Fixes

* add clearData method ([fb63688](https://github.com/raminjafary/vidar/commit/fb63688c3b0ea52839ad8d63f3bfc29fee02d59b))
* add typings for draw options ([884b4fc](https://github.com/raminjafary/vidar/commit/884b4fcf6684c7edf18d1ab7aee3296ae5bd20a6))
* export renderer ([49e5c72](https://github.com/raminjafary/vidar/commit/49e5c72009ca3ef47445180d44aa1897cb43a766))
* set accessor for data ponits ([94b0073](https://github.com/raminjafary/vidar/commit/94b00739929b06fbc254e02a822e112a686e0fa8))
* ts typings ([cdd3ad3](https://github.com/raminjafary/vidar/commit/cdd3ad32be0baa4dca401574565e2ee3707ba171))

## 1.0.0 (2021-06-18)


src/perf.ts
if (typeof module !== 'undefined') module.exports = simpleheat

function simpleheat(this: ThisType<any>, canvas: string): any {
  if (!(this instanceof simpleheat)) return new (simpleheat as any)(canvas)

  this.canvas = canvas = document.getElementById(canvas)

  this.ctx = this.canvas.getContext('2d')
  this.width = canvas.width
  this.height = canvas.height

  this._max = 1
  this._data = []
}

simpleheat.prototype = {
  defaultRadius: 25,
  defaultGradient: {
    0.4: 'blue',
    0.6: 'cyan',
    0.7: 'lime',
    0.8: 'yellow',
    1.0: 'red',
  },
  data(data: any) {
    this._data = data
    return this
  },
  max(max: number) {
    this._max = max
    return this
  },
  add(data: any) {
    this._data.push(data)
    return this
  },
  clear() {
    this._data = []
    return this
  },
  resize() {
    this.width = this.canvas.width
    this.height = this.canvas.height
  },
  createCanvas() {
    if (typeof document != 'undefined') {
      return document.createElement('canvas')
    } else {
      return new this.canvas.constructor()
    }
  },
  radius(r: number, blur: number = 15) {
    const circle: HTMLCanvasElement = (this._circle = this.createCanvas())
    const ctx = circle.getContext('2d')
    const r2 = (this._r = r + blur)

    circle.width = circle.height = r2 * 2

    ctx!.shadowOffsetX = ctx!.shadowOffsetY = r2 * 2
    ctx!.shadowBlur = blur
    ctx!.shadowColor = 'black'

    ctx!.beginPath()
    ctx!.arc(-r2, -r2, r, 0, 2 * Math.PI, true)
    ctx!.closePath()
    ctx!.fill()

    return this
  },
  gradient(grad: any) {
    const canvas: HTMLCanvasElement = this.createCanvas(),
      ctx = canvas.getContext('2d'),
      gradient = ctx!.createLinearGradient(0, 0, 0, 256)

    canvas.width = 1
    canvas.height = 256

    for (const i in grad) {
      gradient.addColorStop(+i, grad[i])
    }

    ctx!.fillStyle = gradient
    ctx?.fillRect(0, 0, 1, 256)

    this._grad = ctx?.getImageData(0, 0, 1, 256).data

    return this
  },

  colorize(pixels: any, gradient: any) {
    for (let i = 0, len = pixels.length, j; i < len; i += 4) {
      j = pixels[i + 3] * 4

      if (j) {
        pixels[i] = gradient[j]
        pixels[i + 1] = gradient[j + 1]
        pixels[i + 2] = gradient[j + 2]
      }
    }
  },
  draw(minOpacity: number) {
    if (!this._circle) this.radius(this.defaultRadius)
    if (!this._grad) this.gradient(this.defaultGradient)

    const ctx = this.ctx

    ctx.clearRect(0, 0, this.width, this.height)

    for (let i = 0, p; i < this._data.length; i++) {
      p = this._data[i]
      ctx.globalAlpha = Math.min(
        Math.max(
          p[2] / this._max,
          minOpacity === undefined ? 0.05 : minOpacity
        ),
        1
      )
      ctx.drawImage(this._circle, p[0] - this._r, p[1] - this._r)
    }

    const data = ctx.getImageData(0, 0, this.width, this.height)
    this.colorize(data.data, this._grad)
    ctx.putImageData(data, 0, 0)

    return this
  },
}
